### PR TITLE
Refactor the Jewellery count plugin into an Item charges plugin

### PR DESCRIPTION
**Pull request by SoyChai**
Completely generalized this plugin's charge-detection code to handle any item which displays a charge count (#) in its name, and added a config panel to choose which categories should be displayed.

image

Potions are hidden by default due to their already-visible liquid  levels.

image

### CODE REVIEW

**Comment by SoyChai** [comment]
Takes care of #1176.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 1]
```diff
@@ -44,15 +46,19 @@
...
-class JewelleryCountOverlay extends Overlay
+class ItemChargesOverlay extends Overlay
 {
+	private final Client client;
+	private final ItemChargesConfig config;
 	private final QueryRunner queryRunner;
 
 	@Inject
-	JewelleryCountOverlay(QueryRunner queryRunner)
+	ItemChargesOverlay(Client client, ItemChargesConfig config, QueryRunner queryRunner)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: make this constructor private

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 2]
```diff
@@ -61,26 +67,27 @@ public Dimension render(Graphics2D graphics)
...
+			Integer charges = ChargeCounts.getCharges(item, config);
 
 			if (charges == null)
 			{
 				continue;
 			}
 
-			final Rectangle bounds = item.getCanvasBounds();
+			final Rectangle bounds = widgetItem.getCanvasBounds();
 			final TextComponent textComponent = new TextComponent();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: Probably saving this object during class initialization and then just altering it and rendering might save some memory.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 3]
```diff
@@ -61,26 +67,27 @@ public Dimension render(Graphics2D graphics)
 	{
 		graphics.setFont(FontManager.getRunescapeSmallFont());
 
-		for (WidgetItem item : getJewelleryWidgetItems())
+		for (WidgetItem widgetItem : getChargeCountWidgetItems())
 		{
-			JewelleryCharges charges = JewelleryCharges.getCharges(item.getId());
+			ItemComposition item = client.getItemDefinition(widgetItem.getId());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: use `final` for variables that will not be modifed as those two

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ChargeCounts.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	 */
+	public static Integer getCharges(ItemComposition item, ItemChargesConfig config)
+	{
+		if (EXCLUDED_ITEMS.contains(item.getId()))
+		{
+			return null;
+		}
+
+		List<String> actions = Arrays.asList(item.getInventoryActions());
+		boolean isMisc = !actions.contains("Rub") && !actions.contains("Drink")
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ChargeCounts.java]: Use final for these variables

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ChargeCounts.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+				&& !UNDRINKABLE_POTIONS.contains(item.getId()) && !BASKETS_AND_SACKS.contains(item.getId());
+
+		if (config.showTeleportCharges() && (actions.contains("Rub") || actions.contains("Activate") || actions.contains("Play"))
+				|| config.showDrinkablePotionDoses() && actions.contains("Drink")
+				|| config.showUndrinkablePotionDoses() && UNDRINKABLE_POTIONS.contains(item.getId())
+				|| config.showBasketAndSackCounts() && BASKETS_AND_SACKS.contains(item.getId())
+				|| config.showMiscItemCharges() && isMisc)
+		{
+			Matcher matcher = Pattern.compile(CHARGES_REGEX).matcher(item.getName());
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ChargeCounts.java]: Use final

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesConfig.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+@ConfigGroup(
+		keyName = "itemCharges",
+		name = "Item Charges",
+		description = "Show charge counts remaining on inventory items"
+)
+public interface ItemChargesConfig extends Config
+{
+	@ConfigItem(
+			keyName = "showTeleportCharges",
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesConfig.java]: This is probably too much indentation, see https://github.com/runelite/runelite/wiki/Code-Conventions.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 7]
```diff
@@ -93,9 +100,9 @@ public Dimension render(Graphics2D graphics)
 		);
 		WidgetItem[] equipmentWidgetItems = queryRunner.runQuery(equipmentQuery);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: Change both inventory and equipment items to final

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 8]
```diff
@@ -93,9 +100,9 @@ public Dimension render(Graphics2D graphics)
...
 
-		Collection<WidgetItem> jewellery = new ArrayList<>();
-		jewellery.addAll(Arrays.asList(inventoryWidgetItems));
-		jewellery.addAll(Arrays.asList(equipmentWidgetItems));
-		return jewellery;
+		Collection<WidgetItem> widgetItems = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: Make this final

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 9]
```diff
@@ -61,26 +67,27 @@ public Dimension render(Graphics2D graphics)
 	{
 		graphics.setFont(FontManager.getRunescapeSmallFont());
 
-		for (WidgetItem item : getJewelleryWidgetItems())
+		for (WidgetItem widgetItem : getChargeCountWidgetItems())
 		{
-			JewelleryCharges charges = JewelleryCharges.getCharges(item.getId());
+			ItemComposition item = client.getItemDefinition(widgetItem.getId());
+			Integer charges = ChargeCounts.getCharges(item, config);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: Final

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java (hunk 2 above)
Comment (Abextm) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: I would rather we just use `graphics.drawString` so we don't have to evaluate the regex too.

**Comment by Kamielvf** [comment]
I hope #1128 still apply to this, nice work.

**Comment by SoyChai** [comment]
I actually want to revisit that issue soon, because it's especially relevant to the topic of adding charge counts to the bank window. Our current digit placement would overlap the stack counts.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java (hunk 2 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: I don't think that is good idea, as one of the purposes of the text component is also consistency of the UI even though that will bring small increase to performance.

File: runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java [hunk 10]
```diff
@@ -61,41 +64,36 @@ public Dimension render(Graphics2D graphics)
...
 
-			final Rectangle bounds = item.getCanvasBounds();
-			final TextComponent textComponent = new TextComponent();
-			textComponent.setPosition(new Point(bounds.x, bounds.y + 16));
-			textComponent.setText(String.valueOf(charges.getCharges()));
-			textComponent.render(graphics);
+			final Rectangle bounds = widgetItem.getCanvasBounds();
+			final int x = bounds.x, y = bounds.y + 16;
+
+			graphics.setColor(Color.BLACK); // Text shadow
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/itemcharges/ItemChargesOverlay.java]: Even though this increases performance by a small bit, this should be changed back to text component as using text component makes it easier to make the UI look consistent and easily changeable in future.

**Comment by Adam-** [comment]
This seems worse than the previous plugin due to the fuzzy () matching + accompanying logic, and harder to follow. What items in addition to the jewelry plugin does this add?

**Comment by SoyChai** [comment]
This version include potions, baskets, sacks, stew spices, waterskins, black masks, castle wars bracelets, abyssal necklaces, ogre bellows, boxed imps, void seals, rods of Ivandis... everything that the old version didn't have hard-coded. The regex also picks up new items by itself assuming Jagex follows their own naming conventions.

I wish the code inside [getCharges()](https://github.com/runelite/runelite/pull/1201/files#diff-bb4b754f66728f45b2c72c5a2329e77cR84) was a little simpler, but I couldn't think of a tidier way to express what it does. The "miscellaneous" category is defined as any item which falls into no other category, so that's why it has a separate check, before the main check which looks at your config settings.

**Comment by jerwuqu** [comment]
I feel like the code would look nicer if it was inclusive rather than exclusive. Then you could also clean `getCharges` up quite a bit by having different item sets.

**Comment by Adam-** [comment]
This was mostly done in #3701

**Pull request closed without merging**